torch/utils/_python_dispatch.py

```diff
@@ -130,17 +130,53 @@ def _pop_mode_temporarily(k: Optional[DispatchKey] = None):
         _push_mode(old, k)
 
 
+@contextlib.contextmanager
+def _temp_pop_torch_function_mode():
+    len = torch._C._len_torch_function_stack()
+    if len > 0:
+        old_value = torch._C._pop_torch_function_stack()
+        try:
+            yield
+        finally:
+            torch._C._push_on_torch_function_stack(old_value)
+    else:
+        yield
+
+
 @contextlib.contextmanager
 def _disable_current_modes():
+    from torch._ops import _len_torch_dispatch_stack_pre_dispatch, _pop_mode_from_pre_dispatch, _set_mode_pre_dispatch
+    from torch._subclasses.functional_tensor import FunctionalTensorMode
+    from torch.fx.experimental.proxy_tensor import ProxyTorchDispatchMode
+    mode_len_pre_dispatch = _len_torch_dispatch_stack_pre_dispatch()
+    old_pre_dispatch_modes = [_pop_mode_from_pre_dispatch() for _ in range(mode_len_pre_dispatch)]
+
+    has_proxy_mode_in_pre_dispatch = False
+    has_functional_mode_in_pre_dispatch = False
+
+    for i in old_pre_dispatch_modes:
+        if isinstance(i, ProxyTorchDispatchMode):
+            has_proxy_mode_in_pre_dispatch = True
+        if isinstance(i, FunctionalTensorMode):
+            has_functional_mode_in_pre_dispatch = True
+
     mode_len = _len_torch_dispatch_stack()
     old_modes = [_pop_mode() for _ in range(mode_len)]
 
+    for old in old_modes:
+        if isinstance(old, FunctionalTensorMode) and has_functional_mode_in_pre_dispatch:
+            raise AssertionError("Can't have FunctionalMode available both in PreDispatch and Python Key")
+        if isinstance(old, ProxyTorchDispatchMode) and has_proxy_mode_in_pre_dispatch:
+            raise AssertionError("Can't have ProxyTorchDispatchMode available both in PreDispatch and Python Key")
+
     # Manually disable proxy and fake modes, if any are active
     try:
-        yield old_modes
+        yield old_pre_dispatch_modes + old_modes
     finally:
         for mode in reversed(old_modes):
             _push_mode(mode)
+        for mode in reversed(old_pre_dispatch_modes):
+            _set_mode_pre_dispatch(mode)
```

Comment:
It feels like we should be able to unify `_push_mode` and `_set_mode_pre_dispatch`, since the mode carries all of the state necessary to know what to do with it (whether it has the pre_dispatch bool tells us which stack it should go on)

Reply:
Will do!